### ISSUE TITLE
Add fault tolerant parsing for `case` expressions

### ISSUE THREAD

**Issue by GearsDatapacks** [gleam-lang/gleam#3978]
Currently, `case something` is a syntax error (obviously). The problem is, the language server cannot provide autocomplete when typing the subject of the case expression. We could make this fault tolerant in the same way as `record.`, to allow for a better LSP experience.

**Comment by giacomocavalieri** [comment]
Love this! It would be a huge dx win

**Comment by lpil** [comment]
Definitely!

**Comment by giacomocavalieri** [comment]
I can think of two ways of approaching this, which one do you think would be best?
1. Syntactically allow writing `case wibble` (with no curly braces) and turn the syntax error into a missing patterns error. Since a case subject is already fault tolerant one would be able to hover over the subject to get its type, trigger code actions, and get completions for free!
2. Keep `case wibble` a syntax error and change the completer to work around the syntax error inspecting the code around the cursor.

To implement 1, one would have to change the parser and nothing much besides that, everything else should already be in place to make it work. I'm not really sure how 2 could be implemented, I haven't looked much at the completer

**Comment by GearsDatapacks** [comment]
1 is probably easier, that's how we handle `wibble.`. I think that's fine unless Louis has any objections.

We would need to be careful though, because some code actions (such as fill missing patterns) rely on the syntax being well formed, and saying that this syntax is valid could lead to odd behaviour there.

**Comment by giacomocavalieri** [comment]
That's great to know! Would you like to work on this or can I give it a try today?

**Comment by GearsDatapacks** [comment]
I'm still away for a the next couple of days, so I won't be able to work on it much until next week. Go for it if you want to!

**Comment by lpil** [comment]
Sounds good to me.

**Comment by GearsDatapacks** [comment]
That's true. We could either do that as part of this, or separately, in preparation for future fault tolerance.

I presume we would do it in a similar way to analysis fault tolerance?

**Comment by giacomocavalieri** [comment]
A small report, I tried going for the easy implementation but as usual the devil's in the details 😆
What happened is it made errors worse in some cases. For example if one writes this:
```gleam
case 1, test {

}
```
`case 1` would now parse correctly as a valid case expression and give an error at `test` with a confusing message.

**Comment by GearsDatapacks** [comment]
Interesting. Surely you could write it in such a way that it would be greedy and parse all subjects before checking for curly braces?

**Comment by giacomocavalieri** [comment]
Yeah I'll try that!
